cmd/rhelemeter-server/main.go

```diff
@@ -0,0 +1,439 @@
+	RequiredLabelFlag []string
+	RequiredLabels    map[string]string
+	Whitelist         []string
+	ElideLabels       []string
+	WhitelistFile     string
+
+	Ratelimit  time.Duration
+	ForwardURL string
+
+	LogLevel string
+	Logger   log.Logger
+
+	TracingServiceName      string
+	TracingEndpoint         string
+	TracingEndpointType     string
+	TracingSamplingFraction float64
+
+	Verbose bool
+}
+
+type Paths struct {
+	Paths []string `json:"paths"`
+}
+
+func (o *Options) Run(ctx context.Context, externalListener, internalListener net.Listener) error {
+	for _, flag := range o.LabelFlag {
+		values := strings.SplitN(flag, "=", 2)
+		if len(values) != 2 {
+			return fmt.Errorf("--label must be of the form key=value: %s", flag)
+		}
+		if o.Labels == nil {
+			o.Labels = make(map[string]string)
+		}
+		o.Labels[values[0]] = values[1]
+	}
+
+	for _, flag := range o.RequiredLabelFlag {
+		values := strings.SplitN(flag, "=", 2)
+		if len(values) != 2 {
+			return fmt.Errorf("--required-label must be of the form key=value: %s", flag)
+		}
+		if o.RequiredLabels == nil {
+			o.RequiredLabels = make(map[string]string)
+		}
+		o.RequiredLabels[values[0]] = values[1]
+	}
+
+	levelledOption := logger.LogLevelFromString(o.LogLevel)
+	o.Logger = level.NewFilter(o.Logger, levelledOption)
+
+	tp, err := tracing.InitTracer(
+		ctx,
+		o.TracingServiceName,
+		o.TracingEndpoint,
+		o.TracingEndpointType,
+		o.TracingSamplingFraction,
+	)
+	if err != nil {
+		return fmt.Errorf("cannot initialize tracer: %v", err)
+	}
+
+	otel.SetErrorHandler(tracing.OtelErrorHandler{Logger: o.Logger})
+
+	var transport http.RoundTripper = otelhttp.NewTransport(&http.Transport{
+		DialContext:         (&net.Dialer{Timeout: 10 * time.Second}).DialContext,
+		MaxIdleConnsPerHost: 10,
+		IdleConnTimeout:     30 * time.Second,
+	})
+
+	if o.Verbose {
+		transport = telemeter_http.NewDebugRoundTripper(o.Logger, transport)
+	}
+
+	forwardClient := &http.Client{
+		Transport: telemeter_http.NewInstrumentedRoundTripper("forward", transport),
+	}
+
+	if o.OIDCIssuer != "" {
+		provider, err := oidc.NewProvider(ctx, o.OIDCIssuer)
+		if err != nil {
+			return fmt.Errorf("OIDC provider initialization failed: %v", err)
+		}
+
+		ctx = context.WithValue(ctx, oauth2.HTTPClient,
+			&http.Client{
+				// Note, that e.g forward timeouts after 5s.
+				Timeout:   20 * time.Second,
+				Transport: telemeter_http.NewInstrumentedRoundTripper("oauth", transport),
+			},
+		)
+
+		cfg := clientcredentials.Config{
+			ClientID:     o.OIDCClientID,
+			ClientSecret: o.OIDCClientSecret,
+			TokenURL:     provider.Endpoint().TokenURL,
+		}
+
+		if o.OIDCAudienceEndpoint != "" {
+			cfg.EndpointParams = url.Values{"audience": []string{o.OIDCAudienceEndpoint}}
+		}
+
+		s := cfg.TokenSource(ctx)
+
+		forwardClient.Transport = &oauth2.Transport{
+			Base:   forwardClient.Transport,
+			Source: s,
+		}
+	}
+
+	var g run.Group
+	{
+		internal := http.NewServeMux()
+
+		// TODO: Refactor to not take *http.Mux
```

Comment:
Why? Are there any problems?

Reply:
This is some copy paste from the telemeter server. Removing it.